hmetrics/init.go

```diff
@@ -55,31 +55,45 @@ func Report(ctx context.Context, pfer printfer) error {
 	if endpoint == "" {
 		return HerokuMetricsURLUnset{}
 	}
-	go report(ctx, endpoint, pfer)
+	if ef == nil {
+		ef = func(_ error) error { return nil }
+	}
+	go report(ctx, endpoint, ef)
 	started = true
 	return nil
 }
 
-func report(ctx context.Context, endpoint string, pfer printfer) {
+// The only thing that should come after and exit() is a return
+func exit() {
+	mu.Lock()
+	defer mu.Unlock()
+	started = false
+}
+
+func report(ctx context.Context, endpoint string, ef ErrHandler) {
 	t := time.NewTicker(metricWaitTime)
 	defer t.Stop()
 
 	for {
 		select {
 		case <-t.C:
 		case <-ctx.Done():
-			mu.Lock()
-			defer mu.Unlock()
-			started = false
+			exit()
```

Comment:
Can this be a single `defer exit()` -- or even inlined again `defer func() { mu.Lock() ...}` -- outside the loop?